core/shared/src/main/scala/cats/effect/IO.scala

```diff
@@ -588,8 +589,21 @@ sealed abstract class IO[+A] private () extends IOPlatform[A] {
   def onCancel(fin: IO[Unit]): IO[A] =
     IO.OnCancel(this, fin)
 
-  def onError(f: Throwable => IO[Unit]): IO[A] =
-    handleErrorWith(t => f(t).voidError *> IO.raiseError(t))
+  @deprecated("Use onError with PartialFunction argument", "3.6.0")
+  def onError(f: Throwable => IO[Unit]): IO[A] = {
+    val pf: PartialFunction[Throwable, IO[Unit]] = { case t => f(t) }
+    onError(pf)
+  }
+
+  /**
+   * Execute a callback on certain errors, then rethrow them. Any non matching error is rethrown
+   * as well.
+   *
+   * Implements `ApplicativeError.onError`.
+   */
+  def onError(pf: PartialFunction[Throwable, IO[Unit]]): IO[A] =
+    handleErrorWith(t =>
+      pf.applyOrElse(t, (_: Throwable) => IO.unit).reportError *> IO.raiseError(t))
```

Comment:
I guess we should use `reportError` for the deprecated `onError`, but for the `onError` consistent with `ApplicativeError` we have to match the semantics defined in Cats.

```suggestion
  @deprecated("Use onError with PartialFunction argument", "3.6.0")
  def onError(f: Throwable => IO[Unit]): IO[A] = {
    val pf: PartialFunction[Throwable, IO[Unit]] = { case t => f(t).reportError }
    onError(pf)
  }

  /**
   * Execute a callback on certain errors, then rethrow them. Any non matching error is rethrown
   * as well.
   *
   * Implements `ApplicativeError.onError`.
   */
  def onError(pf: PartialFunction[Throwable, IO[Unit]]): IO[A] =
    handleErrorWith(t =>
      pf.applyOrElse(t, (_: Throwable) => IO.unit)*> IO.raiseError(t))
```

Reply:
fixed in: https://github.com/typelevel/cats-effect/pull/4121/commits/868451b090e407094d01eaa953ee8b4fde69ac5c